src/main/java/com/google/idea/perf/TracerController.kt

```diff
@@ -183,6 +194,35 @@ class TracerController(
         }
     }
 
+    fun handleCommandChangeFromEdt(rawCmd: String, offset: Int) {
+        val executor = AppExecutorUtil.getAppExecutorService()
+
+        autocompleteFuture?.cancel(true)
+
+        autocompleteFuture = executor.submit(Callable {
+            val result = executor.submit(Callable {
+                autocomplete.predict(rawCmd, offset)
+            })
+
+            try {
+                val suggestions = result.get(AUTOCOMPLETE_MAX_TIMEOUT_MS, MILLISECONDS)
```

Comment:
Also consider using `ProgressIndicatorUtils#withTimeout`.

The advantage of `withTimeout` is that the background process is properly aborted (as long as `ProgressManager#checkCanceled` is called often enough). I.e., it ensures that we don't end up with rogue background process consuming CPU after being canceled.

Similarly, note that `autocompleteFuture?.cancel(true)` has no effect unless `Thread.interrupted()` is called periodically in the background thread. I.e., previously canceled autocomplete invocations may still interact with the UI and cause confusing behavior.

In general it would be nice to enforce an invariant that only one auto-completion session is running at any given moment.